### PR TITLE
[CALCITE-3138] RelStructuredTypeFlattener doesn't restructure ROW type fields

### PR DESCRIPTION
**Pull request by ihuzenko**
PR for [CALCITE-3138](https://issues.apache.org/jira/browse/CALCITE-3138). 

- Fixed restructuring of ROW type
- Removed null indicator because there are cases when it's impossible to track correct new indices using getNewForOldInput(...) method
- Increased depth of flattening for ITEM(...) which gets struct from array by index and the returned struct contains another struct 
- Some ITEM(struct, 'fieldName') expressions are replaced by simple refs after flatten operation
- Added ROW support to SqlItemOperator type inference 
- Increased amount of local variables to simplify debugging 
- Updated tests expected results
- Added check to avoid merging of flatten and restructuring projects

### CODE REVIEW

**Comment by chunweilei** [comment]
Please write the right format of commit message. You can find more details in the doc[1].

[1] https://calcite.apache.org/develop/#contributing.

**Comment by ihuzenko** [comment]
Hi @chunweilei , good catch. Updated commit and PR name.

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java [hunk 1]
```diff
@@ -296,42 +273,42 @@ protected int getNewForOldInput(int oldOrdinal) {
...
-    RelDataType oldInputType = oldInput.getRowType();
-    final int newOffset = calculateFlattenedOffset(oldInputType, oldOrdinal) + existingOffset;
-    newOrdinal += newOffset;
-    final RelDataTypeField field =
-        newInput.getRowType().getFieldList().get(newOffset);
-    return Ord.of(newOrdinal, field.getType());
+    // sum of predecessors post flatten sizes points to new ordinal
+    // of flat field or first field of flattened struct
+    final int postFlatteningOrdinal = currentRel.getInputs().stream()
+        .flatMap(node -> node.getRowType().getFieldList().stream())
```
Comment (vvysotskyi) [inline core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: I'm not sure that it will work for all types of rel nodes. If I understand correctly, there was used assumption, that row type of current rel node consists of fields from the inputs, but for some kinds of operators, like semi-join, fields of one of the input aren't included into the resulting row type.

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java [hunk 2]
```diff
@@ -581,68 +555,52 @@ public void rewriteGeneric(RelNode rel) {
   }
 
   private void flattenProjections(RewriteRexShuttle shuttle,
-      List<? extends RexNode> exps,
-      List<String> fieldNames,
-      String prefix,
-      List<Pair<RexNode, String>> flattenedExps) {
+                                  List<? extends RexNode> exps,
```
Comment (vvysotskyi) [inline core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: Please revert this change, Calcite uses such formatting.

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java [hunk 3]
```diff
@@ -581,68 +555,52 @@ public void rewriteGeneric(RelNode rel) {
...
+      fieldName = prefix + "$" + fieldName;
+    }
+    return fieldName;
+  }
+
   private void flattenProjection(RewriteRexShuttle shuttle,
-      RexNode exp,
-      String fieldName,
-      List<Pair<RexNode, String>> flattenedExps) {
+                                 RexNode exp,
```
Comment (vvysotskyi) [inline core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: And this one.

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java [hunk 4]
```diff
@@ -651,22 +609,57 @@ private void flattenProjection(RewriteRexShuttle shuttle,
...
-                  oldOperands.get(1)));
-        }
-        for (RelDataTypeField field : newExp.getType().getFieldList()) {
-          flattenedExps.add(
-              Pair.of(rexBuilder.makeFieldAccess(newExp, field.getIndex()),
-                  fieldName + "$" + (j++)));
+        List<RexNode> operands = ((RexCall) exp).getOperands();
+
+        RexNode firstOp = operands.get(0);
+        RexNode secondOp = operands.get(1);
```
Comment (vvysotskyi) [inline core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: It may be not safe to obtain the first operand here for the case when we have non-ITEM operator which has single operand.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlItemOperator.java [hunk 5]
```diff
@@ -125,6 +128,14 @@ private SqlSingleOperandTypeChecker getChecker(SqlCallBinding callBinding) {
     case MAP:
       return typeFactory.createTypeWithNullability(operandType.getValueType(),
           true);
+    case ROW:
+      int lastOperandIdx = opBinding.getOperandCount() - 1;
+      String fieldName = opBinding.getOperandLiteralValue(lastOperandIdx, String.class);
+      return operandType.getFieldList().stream()
+          .filter(field -> field.getName().equalsIgnoreCase(fieldName))
+          .findFirst().map(RelDataTypeField::getType)
```
[the remark this answers is not in the document]
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlItemOperator.java]: You can get the `NameMatcher` from `CatalogReader` of `SqlValidator` to decide the case sensitivity.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlItemOperator.java [hunk 6]
```diff
@@ -125,6 +128,14 @@ private SqlSingleOperandTypeChecker getChecker(SqlCallBinding callBinding) {
     case MAP:
       return typeFactory.createTypeWithNullability(operandType.getValueType(),
           true);
+    case ROW:
+      int lastOperandIdx = opBinding.getOperandCount() - 1;
```
[the remark this answers is not in the document]
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlItemOperator.java]: How about the opBinding has 0 operands ?

File: core/src/main/java/org/apache/calcite/sql/fun/SqlItemOperator.java [hunk 7]
```diff
@@ -100,6 +101,8 @@ private SqlSingleOperandTypeChecker getChecker(SqlCallBinding callBinding) {
     case MAP:
       return OperandTypes.family(
           operandType.getKeyType().getSqlTypeName().getFamily());
+    case ROW:
+      return OperandTypes.STRING;
```
[the remark this answers is not in the document]
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlItemOperator.java]: Use `CHARACTER`.

**Comment by danny0405** [comment]
Thanks @ihuzenko , i see there are so many code refactoring which is not relative with the fix in `RelStructuredTypeFlattener`, can we split the code refactoring apart from the main fixing, it's a little hard to figure out what you really want to fix and how the code works when they mixed up.

**Comment by ihuzenko** [comment]
Hi @danny0405 , thanks for participation in review. I divided refactoring and changes into 2 commits, and I'll prepare code review changes commit soon.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlItemOperator.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (vvysotskyi) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlItemOperator.java]: As I pointed in the comment above, for this function allowed only 2 operands: `SqlOperandCountRanges.of(2)`

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java (hunk 1 above)
Comment (ihuzenko) [reply core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: The code is only called when it's exactly was input reference to field returned by previous node. I don't think that problems may be introduced here, because it's basically a little bit more readable version of the previous method.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlItemOperator.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ihuzenko) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlItemOperator.java]: I've used ```getField(...)``` method with case sensitivity ```false``` like in DOT operator.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlItemOperator.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (ihuzenko) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlItemOperator.java]: Thanks for clarification. Fixed.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlItemOperator.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (ihuzenko) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlItemOperator.java]: Done.

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java (hunk 2 above)
Comment (ihuzenko) [reply core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: Fixed.

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java (hunk 3 above)
Comment (ihuzenko) [reply core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: Fixed.

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java (hunk 4 above)
Comment (ihuzenko) [reply core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: Fixed.

**Comment by ihuzenko** [comment]
@vvysotskyi , @danny0405  please take another look at the PR. I made 3 different commits, with changes, refactoring and code review updates.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlItemOperator.java [hunk 8]
```diff
@@ -125,6 +128,14 @@ private SqlSingleOperandTypeChecker getChecker(SqlCallBinding callBinding) {
     case MAP:
       return typeFactory.createTypeWithNullability(operandType.getValueType(),
           true);
+    case ROW:
+      int lastOperandIdx = opBinding.getOperandCount() - 1;
+      String fieldName = opBinding.getOperandLiteralValue(lastOperandIdx, String.class);
+      return operandType.getFieldList().stream()
```
[the remark this answers is not in the document]
Comment (ihuzenko) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlItemOperator.java]: done, thanks.

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java [hunk 9]
```diff
@@ -587,86 +561,103 @@ private void flattenProjections(RewriteRexShuttle shuttle,
...
+                from += postFlattenSize(field.getType());
+              }
+            }
+          } else if (firstOp instanceof RexCall) {
+            // to get nested struct from return type of firstOp rex call,
+            // we need to flatten firstOp and get range of expressions which
+            // corresponding to desirable nested struct flattened fields
+            List<Pair<RexNode, String>> firstOpFlattenedExps = new ArrayList<>();
+            flattenProjection(shuttle, firstOp, fieldName, firstOpFlattenedExps);
+            int newInnerOrdinal = getNewInnerOrdinal(firstOp, literalString);
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: It does not seem right we still passed in the `fieldName ` which is the name of the `exp` but not the `firstOp` which does not have a explicit name.

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java [hunk 10]
```diff
@@ -934,7 +968,21 @@ private RexNode flattenComparison(
...
       }
     }
+
+  }
+
+  private int getNewInnerOrdinal(RexNode firstOp, String literalString) {
+    int newInnerOrdinal = 0;
+    for (RelDataTypeField field : firstOp.getType().getFieldList()) {
+      if (literalString.equalsIgnoreCase(field.getName())) {
+        break;
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: I don't think this is right to match the fist flattened field just by name, how do you know what the (struct type field original) operator is ? It seems only right if they are all `ITEM` operators.

Comment (ihuzenko) [reply core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: Yes, the method only used for item with string second operand.

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java (hunk 9 above)
Comment (ihuzenko) [reply core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: I think passing old name here is better for preserving some history, although if you have reasonable alternative, I won't mind changing the code.

**Comment by ihuzenko** [comment]
Hello @danny0405 , I've answered your last comments. Should I do some changes or PR can be merged ?

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java (hunk 9 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: We should not go into the recursive loop again with a wrong name `fieldName`, this name is used as a look up key( case insensitive) for `ITEM` operator.

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java (hunk 10 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: That means this code snippet only works for full `ITEM` operator, but there is no any code can guarantee this, the most weird part is that you match the `first` field with name, that confused me a lot, why not the second ? the third ?

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java (hunk 9 above)
Comment (ihuzenko) [reply core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: Here the ```fieldName``` has meaning of result name of projection expression, when it's necessary to lookup field within struct (ITEM operator), then used ```literalString``` (value of ITEM's second operand).

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java (hunk 10 above)
Comment (ihuzenko) [reply core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: 1) Every call to the method is performed only after the check. 
```
      if (rexCall.op == SqlStdOperatorTable.ITEM
          && rexCall.operands.get(0).getType().isStruct()
          && rexCall.operands.get(1).isA(SqlKind.LITERAL)
          && SqlTypeUtil.inCharFamily(rexCall.operands.get(1).getType())) {
```
So I think it's enough to guarantee that we're looking for inner field within struct returned by first operand of ITEM.   

2) What's actually wrong with matching of first field if iteration is performed for non flattened struct fields ? There is no threat that some nested field with same name will come up and match before the desired one. Maybe I misunderstand you and you're just worried about case insensitive match, then in my opinion it's unlikely that users will create structs, like ```STRUCT <aaa:INT, AAA:STRING>``` .

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: 1. You indeed do the check:
```java
      if (rexCall.op == SqlStdOperatorTable.ITEM
          && rexCall.operands.get(0).getType().isStruct()
          && rexCall.operands.get(1).isA(SqlKind.LITERAL)
          && SqlTypeUtil.inCharFamily(rexCall.operands.get(1).getType())) {
```
before calling this method, but within the method, you still have a recursive logic to flatten all the struct fields, which i think has already exceed the limitation of the check.

2. Assume we have an expression a(b(c, 'k1'), 'k2')['k1'] (a, b, c all return struct types), is it really right we return the field of field c['k1'] ? And how do you know that a, b, and c are all `ITEM` operators ?

Comment (ihuzenko) [reply core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: 1. There is no further recursive calls within the ```getNewInnerOrdinal(RexNode firstOp, String literalString)``` method. Only recursion is applied to predict post flatten size of field which resides before desired one, and this ```postFlattenSize(field.getType())``` prediction doesn't require any matching.

Comment (ihuzenko) [reply core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: 2. In the absence of recursive search for non item operands, the method will look for ```'k1'``` inside result type of ```a(b(c, 'k1'), 'k2')```, note that type is not flattened at the moment of search.

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: The method `getNewInnerOrdinal` is invoked after the recursive invocation `flattenProjection`, why do you think the type is not flattened ?

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java (hunk 9 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: Maybe we should use a name pattern like `fieldName$0`, if it was in a deep recursive, the name may like `fieldName$0$0...`

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java (hunk 10 above)
Comment (ihuzenko) [reply core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: Recursive ```flattenProjection``` call in the case will just fill list ```firstOpFlattenedExps``` and don't change state of object ```firstOp``` .

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java (hunk 9 above)
Comment (ihuzenko) [reply core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: ok, I'll add the concatenation.

Comment (ihuzenko) [reply core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: @danny0405 done.

**Comment by ihuzenko** [comment]
@vvysotskyi , @danny0405 thanks for review. I've squashed the commits.
